home/rvagg/2021-02-exploration-report-small-filecoin-deals.md

```diff
@@ -0,0 +1,134 @@
+ * Used minimal stateroots snapshot @ https://fil-chain-snapshots-fallback.s3.amazonaws.com/mainnet/minimal_finality_stateroots_latest.car
+ * Started manually with `lotus daemon --import-snapshot`
+ * Stopped, then started via systemd, let it sync, watching the log output with `tail -f`
+ * Can interact with the daemon via the CLI with `lotus client` and `lotus wallet`, etc. as the `lotus` user (appears to be user-restricted IPC)
+ * Made a wallet and transferred some FIL in it for experimenting
+
+Interesting commands:
+
+ * `lotus sync wait` will run until the chain is synced, zero exit status on success
+ * `lotus wallet` - `lotus wallet new`, `lotus wallet new`, `lotus wallet balance`, `lotus wallet export` (export privkey for backup)
+ * `lotus client`
+   * `lotus client import <path>` to import file(s) into the client (somewhere?), presumably as UnixFS (?), gets a CID
+   * `lotus client local` shows you what you have imported and their CIDs
+   * `lotus client balances` shows you funds that you have active in the "market" (where you have to have funds in order to make deals etc. i.e. out of your wallet)
+   * `lotus client find <CID>` lets you query the network for a CID
+
+Deals:
+
+   * `lotus client query-ask <minerID>` ask a miner for their deal terms - returns FIL asking price per GiB per epoch (30 seconds) for both regular and verified deals, along with minimum and maximum deal sizes.
+   * `lotus client list-asks` will do a scan of the network for miners with power and do a `query-ask` on each of them, reporting the output to stdout, one miner per line (~450)
+   * `lotus client deal` without additional arguments takes you through a prompted deal-making process that asks for your data CID, a miner ID, whether it's a verified deal; it queries the miner, asks you to confirm the asking price and starts the deal process if you say `yes`.
+   * `lotus client list-deals` to show you deals in progress and their status (Note: the status identifiers are not self-explainatory, the best source of documentation seems to be the markets code: https://github.com/filecoin-project/go-fil-markets/blob/master/storagemarket/dealstatus.go)
+   * `lotus client deal-stats` seems to be the best way to view the outcome of your deals once they disappear off `list-deals`
+   * The Lotus daemon log also has output related to deals, you can grep for the status, e.g. `grep StorageDealError ~/var/lotus-daemon.log` (mostly just `"error reading Response message: EOF"`)
+
+### Making deals
+
+* Imported a 9.6 Mb (`10016091` b) JPG file to Lotus with `lotus client import`. Some notes about sizes:
+  * Note that data is _padded_ when making deals. First 2 empty bits per 254 bytes are added (called "fr32 padding", `Math.ceil(10016091 * (256/254))`), then it's rounded up to the next nearest power-of-two value as "pieces" stored in Filecoin have to fit neatly into a proving tree (`2**Math.ceil(Math.log2(Math.ceil(10016091*(256/254))))`).
```

Comment:
There's a wolfram snippet circulating in the community for this: https://www.wolframalpha.com/input/?i=x+%3D+10016091%3B+128*%28+2%5E%28+ceil%28+log2%28+ceil+%28+x+%2F127+%29+%29+%29+%29+%29

Reply:
yeah, saw it, and I hate it, so I'm not mentioning it

Reply:
fyi there's a further conversation related to this piece of the documentation: https://github.com/filecoin-project/filecoin-docs/issues/705